### PR TITLE
Properly handle lisp quotes (' symbol) and numbers starting with '+'

### LINKED ISSUE #105: Can't parse symbols according to spec. More powerful "parse_edn" pattern matching?
According to https://github.com/edn-format/edn#symbols the following is valid (and confirmed with clojure.edn/read-string).

```
(clojure.edn/read-string "-foobar")
-foobar
(type (clojure.edn/read-string "-foobar"))
clojure.lang.Symbol
```

Currently parse_edn is only matching on the first character, which is insufficient for this case. `-foobar` gets treated as a Number.
https://github.com/naomijub/edn-rs/blob/8353fc4cbfb1938849ffed0a4a53951b95706f24/src/deserialize/parse.rs#L41

I believe it should be possible to be passing around a `&str` instead of `&mut std::iter::Enumerate<std::str::Chars>`, which would allow for str pattern matching. I don't know if this is a slippery slope eventually recreating regex. Before a big change, a discussion doesn't seem like a bad idea. There might be an even easier way or additional considerations.

Rational:
1. New edge case 
2. It could help clean up #103 
3. If this was done, all the very expensive and heap allocating clone()s on the Enumerate could be removed.
4. None of this is externally pub, so no other projects would break. I believe it would technically be possible to even have the return not be an owned String but str slices from the original &str too. This would make it so there's no heap allocations, but would also break existing code. Perhaps worth considering before 1.0.0.

### PR DESCRIPTION
**Pull request by Grinkers**
Fixes #105

### CODE REVIEW

**Comment by Grinkers** [comment]
@naomijub 

While doing this, I also ran into
https://github.com/Grinkers/edn-rs/blob/symbols/src/deserialize/parse.rs#L69
which I believe should be `(|c| !c.1.is_whitespace() && !DELIMITERS.contains(&c.1))`

when doing tests, I ran into 
```
---- deserialize::test::namespaced_maps_navigation stdout ----
thread 'deserialize::test::namespaced_maps_navigation' panicked at 'assertion failed: `(left == right)`
  left: `Nil`,
 right: `Key(":val")`', src/deserialize/mod.rs:665:9
 
 ---- deserialize::test::namespaced_maps stdout ----
thread 'deserialize::test::namespaced_maps' panicked at 'assertion failed: `(left == right)`
  left: `Key(":abc")`,
 right: `NamespacedMap("abc", Map({"0": Key(":val"), "1": Key(":value")}))`', src/deserialize/mod.rs:572:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
I can't find any reference to namespaced maps in EDN (or clojure, for that matter). Is this some sort of reader macro?

```
(clojure.edn/read-string ":abc{ 0 :val 1 :value}")
:abc
```
is what I would have expected and get.

File: src/deserialize/parse.rs [hunk 1]
```diff
@@ -1,7 +1,7 @@
 use crate::edn::{Edn, Error, List, Map, Set, Vector};
 use std::collections::{BTreeMap, BTreeSet};
 
-const DELIMITERS: [char; 5] = [',', ']', '}', ')', ';'];
+const DELIMITERS: [char; 8] = [',', ']', '}', ')', ';', '(', '[', '{'];
```
Comment (naomijub) [inline src/deserialize/parse.rs:4]: might be missing set delimiter, but we have a workaround to deal with sets, by replacing the `#{` with `@`

File: src/deserialize/parse.rs [hunk 2]
```diff
@@ -43,10 +43,12 @@ pub fn parse_edn(
     match c {
         Some((_, '\"')) => read_str(chars),
         Some((_, ':')) => read_key_or_nsmap(chars),
-        Some((_, '-')) => Ok(read_number('-', chars)?),
+        Some((_, n)) if n.is_numeric() => Ok(read_number(n, chars)?),
+        Some((_, n)) if (n == '-' || n == '+') && read_number(n, &mut chars.clone()).is_ok() => {
```
Comment (naomijub) [inline src/deserialize/parse.rs]: Maybe this ` read_number(n, &mut chars.clone()).is_ok()` could be a `.peek().next().is_numeric()`?

Comment (naomijub) [reply src/deserialize/parse.rs]: Also, might be better to do  `Some((_, '+') | (_, '-')) if ...` .


Might be `Some((_, '+')) | Some((_, '-')) if ...` , not sure

**Comment by naomijub** [comment]
@Grinkers 
Our implementation of namespace mas is based off https://github.com/edn-format/edn#tagged-elements. It was adapting to Nubank's coding style at the time

**Comment by naomijub** [comment]
@Grinkers I will look this error later today and get back to you

File: src/deserialize/parse.rs (hunk 2 above)
Comment (Grinkers) [reply src/deserialize/parse.rs]: > 

The 'n' needs to be passed on for later processing of - or + (new edge case mentioned in the edn spec), so can't just be matched and discarded.

Comment (Grinkers) [reply src/deserialize/parse.rs]: > Maybe this ` read_number(n, &mut chars.clone()).is_ok()` could be a `.peek().next().is_numeric()`?

Nice catch.

**Comment by naomijub** [comment]
I'm really sorry for the time we took to review. We were both sick this past month

**Pull request merged**